### ISSUE TITLE
Unspecified dimension after tf.sparse_tensor_dense_matmul

### ISSUE THREAD

**Issue by rueberger** [tensorflow/tensorflow#4282]
### Environment info

Operating System:

Problem encountered on linux CPU build installed from `47501a5ebc62fcb8a3d7832722d39997696897dc`.  
### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

```
import tensorflow as tf
indices = [[0, 0], [1, 2]]
values = [1., 1.]
shape = [4, 6]
test_sparse = tf.SparseTensor(indices, values, shape)
v = tf.ones((6, 1))

sess = tf.Session()
sess.run(tf.initialize_all_variables())

mult = tf.sparse_tensor_dense_matmul(test_sparse, v)
print(mult)
--> <tf.Tensor 'SparseTensorDenseMatMul_1/SparseTensorDenseMatMul:0' shape=(?, 1) dtype=float32>

# but of course
sess.run(mult).shape
--> (4, 1)
```

My question is: why is it that the shape of `mult` is partially unspecified before run time? It can be inferred from the shape of `test_sparse` and `v` that `mult` will have shape `(4, 1)`. 

This is very annoying because it prevents me from initializing a Variable with the result of a sparse matmul as in the following:

```
mult_val = tf.Variable(tf.sparse_tensor_dense_matmul(test_sparse, v))
--> ValueError: initial_value must have a shape specified: Tensor("SparseTensorDenseMatMul/SparseTensorDenseMatMul:0", shape=(?, 1), dtype=float32)
```

On the other hand, dense matmul has the desired behavior:

```
dense_mult = tf.matmul(tf.sparse_tensor_to_dense(test_sparse, v))
dense_mult.get_shape()
--> TensorShape([Dimension(4), Dimension(1)])
```

So why does sparse matmul behave differently? Is this a bug? An implementation quirk? 

In the meanwhile, how can I initialize a Variable with the result of a sparse matmul without having to densify the sparse matrix first?

**Comment by rueberger** [comment]
OK, there is an easy solution to this problem. Simply construct the variable with

```
mult_val = tf.Variable(tf.sparse_tensor_dense_matmul(test_sparse, v), validate_shape=False)
```

I'm leaving this open though, because I still don't see why sparse_tensor_dense_matmul should return a tensor with unspecified shape.

**Comment by tatatodd** [comment]
@rueberger nice that you found as solution to your immediate problem!
@concretevitamin can correct me if what I say below is wrong, or if we have future plans to improve things.

I think you can classify this as an "implementation quirk".  Let's start with the representation of sparse tensors.  Basically you can think of SparseTensor as a convenient wrapper around 3 underlying tensors, which represent the indices, values and shape independently:
https://www.tensorflow.org/versions/r0.10/api_docs/python/sparse_ops.html#SparseTensor

For a regular ops on non-sparse tensors, the shape of the output is determined by the shapes of the input tensors, as you expect.

Let's say that's all we're given, and we try to implement SparseTensor ourselves, on top of this framework.  Well the problem is that the shape of the SparseTensor is no longer fixed; since it is itself represented as a tensor.

Another way to look at at:
* The shape of `[1,2,3]` is `[3]`, since we have a 1-dimensional vector with 3 elements.
* The shape of `[4,6]` is `[2]`
* The shape of the `shape = [4, 6]` variable in your example is `[2]`

So basically if we limit ourselves to looking at the shapes, and not at the values, we end up the behavior you've reported.  Hope this helps!

**Comment by rueberger** [comment]
Ah, of course, thanks for the explanation. Perhaps it would be possible to slightly modify tf.sparse_tensor_dense_matmul to have a defined shape when the `shape` argument is passed as an array? That would be a pretty easy change, no?

**Comment by drpngx** [comment]
Is this still current?

**Comment by davharris** [comment]
Possibly. I just reported crashing behavior that seems to be related to the dimension of the tensors in sparse_tensor_dense_matmul (#7417). I'm on version 0.12.1.

**Comment by davharris** [comment]
(I'm mainly referring to "crash 2" from that report, where `sparse_tensor_dense_matmul_op.cc` seems to get confused about the number of rows.  I'll let you know when I have a reproducible example).

**Comment by davharris** [comment]
(example added.  Sorry for the triple comment!)

**Comment by girving** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
